lib/llvmopencl/SubCfgFormation.cc

```diff
@@ -0,0 +1,1547 @@
+  for (auto *BB : Blocks_) {
+    auto *NewBB = llvm::CloneBasicBlock(
+        BB, VMap, ".subcfg." + llvm::Twine{EntryId_} + "b", &F);
+    VMap[BB] = NewBB;
+    NewBlocks_.push_back(NewBB);
+    for (auto *Succ : llvm::successors(BB)) {
+      if (auto ExitIt = ExitIds_.find(Succ); ExitIt != ExitIds_.end()) {
+        NewBlocks_.push_back(createExitWithID(*ExitIt, NewBB, AfterBB));
+      }
+    }
+  }
+
+  LoadBB_ = createLoadBB(VMap);
+
+  VMap[EntryBarrier_] = LoadBB_;
+
+  llvm::SmallVector<llvm::BasicBlock *, 3> Latches;
+  llvm::BasicBlock *LastHeader = LoadBB_;
+  llvm::Value *Idx = ContIdx_;
+
+  createLoopsAround(F, AfterBB, LocalSize, EntryId_, VMap, Latches, LastHeader,
+                    Idx);
+
+  PreHeader_ = createUniformLoadBB(LastHeader);
+  LastHeader->replacePhiUsesWith(&F.getEntryBlock(), PreHeader_);
+
+  print();
+
+  addRemappedDenseMapKeys(InstAllocaMap, VMap, RemappedInstAllocaMap);
+  loadMultiSubCfgValues(InstAllocaMap, BaseInstAllocaMap, ContInstReplicaMap,
+                        PreHeader_, VMap);
+  loadUniformAndRecalcContValues(BaseInstAllocaMap, ContInstReplicaMap,
+                                 PreHeader_, VMap);
+
+  llvm::SmallVector<llvm::BasicBlock *, 8> BlocksToRemap{NewBlocks_.begin(),
+                                                         NewBlocks_.end()};
+  llvm::remapInstructionsInBlocks(BlocksToRemap, VMap);
+
+  removeDeadPhiBlocks(BlocksToRemap);
+
+  EntryBB_ = PreHeader_;
+  ExitBB_ = Latches[0];
+  ContIdx_ = Idx;
+}
+
+// remove incoming PHI blocks that no longer actually have an edge to the PHI
+void SubCFG::removeDeadPhiBlocks(
+    llvm::SmallVector<llvm::BasicBlock *, 8> &BlocksToRemap) const {
+  for (auto *BB : BlocksToRemap) {
+    llvm::SmallPtrSet<llvm::BasicBlock *, 4> Predecessors{llvm::pred_begin(BB),
+                                                          llvm::pred_end(BB)};
+    for (auto &I : *BB) {
+      if (auto *Phi = llvm::dyn_cast<llvm::PHINode>(&I)) {
+        llvm::SmallVector<llvm::BasicBlock *, 4> IncomingBlocksToRemove;
+        for (int IncomingIdx = 0; IncomingIdx < Phi->getNumIncomingValues();
+             ++IncomingIdx) {
+          auto *IncomingBB = Phi->getIncomingBlock(IncomingIdx);
+          if (!Predecessors.contains(IncomingBB))
+            IncomingBlocksToRemove.push_back(IncomingBB);
+        }
+        for (auto *IncomingBB : IncomingBlocksToRemove) {
+          llvm::errs() << "[SubCFG] Remove incoming block "
+                       << IncomingBB->getName() << " from PHI " << *Phi << "\n";
+          Phi->removeIncomingValue(IncomingBB);
+          llvm::errs() << "[SubCFG] Removed incoming block "
+                       << IncomingBB->getName() << " from PHI " << *Phi << "\n";
+        }
+      }
+    }
+  }
+}
+
+// // check if a contiguous value can be tracked back to only uniform values and
+// // the wi-loop indvar currently cannot track back the value through PHI
+// nodes. bool dontArrayifyContiguousValues(
```

Comment:
Can you comment out with #if 0 if you plan to use this code in the future and the reason why it's included in the repo?